### PR TITLE
New module: Basicpy

### LINKED ISSUE #3089: new module: BaSicPy
### Is there an existing module for this?

- [x] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by FloWuenne**
## PR checklist

Closes #3089  

This PR adds the new module Basicpy for illumination correction of images using python.

**test-datasets**: This module uses existing test data from the `imaging` folder.

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by FloWuenne** [comment]
One of these cases again where I don't understand why the md5sum is not the same in the Action tests. In my gitpod environment, md5sums are reproducible!

File: modules/nf-core/basicpy/main.nf [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    container "yfukai/basicpy-docker-mcmicro:0.1.2"
+
+    input:
+    tuple val(meta), path(image)
+    val(cpu_gpu)
+
+    output:
+    tuple val(meta), path("*.tiff"), emit: fields
+    path "versions.yml"           , emit: versions
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/main.nf:18]:
```suggestion
    tuple val(meta), path("*.tiff"), emit: fields
    path "versions.yml"            , emit: versions
```

File: modules/nf-core/basicpy/main.nf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    def VERSION = "0.1.2" // WARN: Version information not provided by tool on CLI. Please update this string when bumping
+    """
+    /opt/main.py $cpu_gpu $image . $args
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        basicpy:: $VERSION
+    END_VERSIONS
+    """
+}
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/main.nf:35]: You can probably guess this already, but can you add some stub? 😁

File: modules/nf-core/basicpy/meta.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  - "basicpy":
+      description: "BaSiCPy is a python package for background and shading correction of optical microscopy images. It is developed based on the Matlab version of BaSiC tool with major improvements in the algorithm."
+      homepage: "https://github.com/peng-lab/BaSiCPy"
+      documentation: "https://basicpy.readthedocs.io/en/latest/index.html"
+      tool_dev_url: "https://github.com/peng-lab/BaSiCPy"
+      doi: "doi: 10.1038/ncomms14836."
+      licence: "MIT License"
+
+input:
+  # Only when we have meta
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/meta.yml:15]:
```suggestion
```

File: modules/nf-core/basicpy/meta.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+name: "basicpy"
+description: write your description here
+keywords: illumiation_correction, microscopy, imaging
+  - sort
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/meta.yml:4]: Can you add some more keywords?

File: modules/nf-core/basicpy/meta.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test' ]
+
+  - image:
+      type: file
+      description: Tiff file to be used for dark and flat field illumination correction
+      pattern: "*.{tiff,tif}"
+
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/meta.yml:26]: `cpu_gpu` is missing :)

File: modules/nf-core/basicpy/meta.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+output:
+  #Only when we have meta
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/meta.yml:28]:
```suggestion
```

File: modules/nf-core/basicpy/meta.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test' ]
+  #
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/meta.yml:34]:
```suggestion
```

File: modules/nf-core/basicpy/meta.yml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+
+  - fields:
+      type: Tiff fields for dark and flat field illumination correction
+      description:
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/meta.yml:42]: It seems you typed the description in the type field 😁

File: modules/nf-core/basicpy/main.nf [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    // Exit if running this module with -profile conda / -profile mamba
+    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
+        exit 1, "Basicpy module does not support Conda. Please use Docker / Singularity instead."
+    }
+
+    container "yfukai/basicpy-docker-mcmicro:0.1.2"
+
+    input:
+    tuple val(meta), path(image)
+    val(cpu_gpu)
```
Comment (nvnieuwk) [inline modules/nf-core/basicpy/main.nf:14]: Is this mandatory?

File: tests/modules/nf-core/basicpy/nextflow.config [hunk 10]
```diff
@@ -0,0 +1,7 @@
+process {
+
+    publishDir = { "${params.outdir}/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" }
+
+}
+
+docker.runOptions = '--entrypoint ""'
```
Comment (nvnieuwk) [inline tests/modules/nf-core/basicpy/nextflow.config:7]: Do you need this option to run the module? or is it only for tests?

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Some small feedback but overall good job!

File: modules/nf-core/basicpy/main.nf (hunk 9 above)
Comment (FloWuenne) [reply modules/nf-core/basicpy/main.nf:14]: Yes, chosing cpu or gpu is a mandatory positional argument for how this container is built. I will add this note into the meta.yml!

File: modules/nf-core/basicpy/meta.yml (hunk 4 above)
Comment (FloWuenne) [reply modules/nf-core/basicpy/meta.yml:4]: I think I just added the keywords wrong as one line, rather then bullet points! Changed this!

File: tests/modules/nf-core/basicpy/nextflow.config (hunk 10 above)
Comment (FloWuenne) [reply tests/modules/nf-core/basicpy/nextflow.config:7]: Due to the Entrypoint of the container applying directly the python script, this was the only way I could make it work, so I would. think this is required to run the docker version.

File: modules/nf-core/basicpy/meta.yml (hunk 5 above)
Comment (FloWuenne) [reply modules/nf-core/basicpy/meta.yml:26]: I added it here. What is the proper way to have a mandatory input parameter though? Documentation for modules says no values as MANDATORY input channels...

File: modules/nf-core/basicpy/main.nf (hunk 2 above)
Comment (FloWuenne) [reply modules/nf-core/basicpy/main.nf:46]: I have attempted to add a basic stub here. Please correct anything that's wrong!

**Comment by FloWuenne** [comment]
Still have the issue where the md5sum is not the same here in the Action then in gitpod... Any ideas why @nvnieuwk ? Inside Gitpod its always reproducible I think.

File: modules/nf-core/basicpy/main.nf (hunk 9 above)
Comment (nvnieuwk) [reply modules/nf-core/basicpy/main.nf:14]: Perfect! Thanks!

File: tests/modules/nf-core/basicpy/nextflow.config (hunk 10 above)
Comment (nvnieuwk) [reply tests/modules/nf-core/basicpy/nextflow.config:7]: Maybe this should also be mentioned in the meta.yml then :)

**Comment by nvnieuwk** [comment]
> Still have the issue where the md5sum is not the same here in the Action then in gitpod... Any ideas why @nvnieuwk ? Inside Gitpod its always reproducible I think.

I'm not sure... Could be a lot of things but mainly it's the presence of a date in the file (although I'm not sure why it would only fail here...)

File: tests/modules/nf-core/basicpy/nextflow.config (hunk 10 above)
Comment (FloWuenne) [reply tests/modules/nf-core/basicpy/nextflow.config:7]: Where would I add this to the meta.yml though? There isn't really any space for comments is there? I tried adding one to the tool and it caused failing of the lint. Feel free to add to finalize this :) !

Comment (nvnieuwk) [reply tests/modules/nf-core/basicpy/nextflow.config:7]: Maybe you can add this to the main description of the tool, it's not ideal but I don't think there is a better place available atm

**Comment by FloWuenne** [comment]
@nvnieuwk I don't understand why the pytest is failing due to versions.yml now. Is it because of the stub I added? Again, did not fail this test in the Gitpod environment... Any clues?

**Comment by nvnieuwk** [comment]
When I run on Gitpod it exits with exit code 130. That's reason tests are failing. I'm not sure what could be the cause for this, but maybe you do?

**Comment by FloWuenne** [comment]
Found the issue. New docker container was pushed to dockerhub but with same version number. So the Github actions pulled the new container with changed CLI code, while my Gitpod was using the old container...

**Comment by nvnieuwk** [comment]
You'll probably need to add a `b` or something after the version to make sure it has a unique version number

**Comment by FloWuenne** [comment]
@nvnieuwk Seems like it passes all the tests now! Can you do a final check and let me know if aything is missing please :)?

**Pull request merged**